read_sql does not accept SqlAlchemy ``Union[Select, FromStatement]`` and ``Connection``

**Describe the bug**
A clear and concise description of what the bug is.

Passing a connection based on a Session to SQLAlchemy is not supported

**To Reproduce**
1. Provide a minimal runnable `pandas` example that is not properly checked by the stubs.
2. Indicate which type checker you are using (`mypy` or  `pyright`).
3. Show the error message received from that type checker while checking your example.

```
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

engine = create_engine('postgresql://Something')

Session = sessionmaker(engine)

with Session() as session:
    pd.read_sql(session.query("Something").statement, session.connection())
```

This returns:

```
error: No overload variant of "read_sql" matches argument types "Union[Select, FromStatement]", "Connection"  [call-overload]
note: Possible overload variants:
note:     def read_sql(sql: str, con: Union[str, Connectable, Connection], index_col: Union[str, List[str], None] = ..., coerce_float: bool = ..., params: Union[List[str], Tuple[str, ...], Dict[str, str], None] = ..., parse_dates: Union[List[str], Dict[str, str], Dict[str, Dict[str, Any]], None] = ..., columns: List[str] = ..., *, chunksize: int) -> Generator[DataFrame, None, None]
note:     def read_sql(sql: str, con: Union[str, Connectable, Connection], index_col: Union[str, List[str], None] = ..., coerce_float: bool = ..., params: Union[List[str], Tuple[str, ...], Dict[str, str], None] = ..., parse_dates: Union[List[str], Dict[str, str], Dict[str, Dict[str, Any]], None] = ..., columns: List[str] = ..., chunksize: None = ...) -> DataFrame
```

**Please complete the following information:**
 - OS: [e.g. Windows, Linux, MacOS] MacOS
 - OS Version [e.g. 22] 12.5.1
 - python version 3.10
 - version of type checker mypy 0.960
 - version of installed `pandas-stubs` [v1.4.4.220912]


**Additional context**
Add any other context about the problem here.

--- a/pandas-stubs/io/sql.pyi
+++ b/pandas-stubs/io/sql.pyi
@@ -5,12 +5,14 @@ from typing import (
     Generator,
     Iterable,
     Literal,
+    Union,
     overload,
 )
 
 from pandas.core.base import PandasObject
 from pandas.core.frame import DataFrame
 import sqlalchemy.engine
+import sqlalchemy.sql.expression
 
 from pandas._typing import (
     DtypeArg,
@@ -20,10 +22,16 @@ from pandas._typing import (
 # TODO: Remove after switch to 1.5.x, moved to pandas.errors
 class DatabaseError(IOError): ...
 
+_SQLConnection = Union[
+    str,
+    sqlalchemy.engine.Connectable,
+    sqlite3.Connection,
+]
+
 @overload
 def read_sql_table(
     table_name: str,
-    con: str | sqlalchemy.engine.Connectable | sqlite3.Connection,
+    con: _SQLConnection,
     schema: str | None = ...,
     index_col: str | list[str] | None = ...,
     coerce_float: bool = ...,
@@ -35,7 +43,7 @@ def read_sql_table(
 @overload
 def read_sql_table(
     table_name: str,
-    con: str | sqlalchemy.engine.Connectable | sqlite3.Connection,
+    con: _SQLConnection,
     schema: str | None = ...,
     index_col: str | list[str] | None = ...,
     coerce_float: bool = ...,
@@ -45,8 +53,8 @@ def read_sql_table(
 ) -> DataFrame: ...
 @overload
 def read_sql_query(
-    sql: str,
-    con: str | sqlalchemy.engine.Connectable | sqlite3.Connection,
+    sql: str | sqlalchemy.sql.expression.Selectable,
+    con: _SQLConnection,
     index_col: str | list[str] | None = ...,
     coerce_float: bool = ...,
     params: list[str] | tuple[str, ...] | dict[str, str] | None = ...,
@@ -57,8 +65,8 @@ def read_sql_query(
 ) -> Generator[DataFrame, None, None]: ...
 @overload
 def read_sql_query(
-    sql: str,
-    con: str | sqlalchemy.engine.Connectable | sqlite3.Connection,
+    sql: str | sqlalchemy.sql.expression.Selectable,
+    con: _SQLConnection,
     index_col: str | list[str] | None = ...,
     coerce_float: bool = ...,
     params: list[str] | tuple[str, ...] | dict[str, str] | None = ...,
@@ -68,8 +76,8 @@ def read_sql_query(
 ) -> DataFrame: ...
 @overload
 def read_sql(
-    sql: str,
-    con: str | sqlalchemy.engine.Connectable | sqlite3.Connection,
+    sql: str | sqlalchemy.sql.expression.Selectable,
+    con: _SQLConnection,
     index_col: str | list[str] | None = ...,
     coerce_float: bool = ...,
     params: list[str] | tuple[str, ...] | dict[str, str] | None = ...,
@@ -80,8 +88,8 @@ def read_sql(
 ) -> Generator[DataFrame, None, None]: ...
 @overload
 def read_sql(
-    sql: str,
-    con: str | sqlalchemy.engine.Connectable | sqlite3.Connection,
+    sql: str | sqlalchemy.sql.expression.Selectable,
+    con: _SQLConnection,
     index_col: str | list[str] | None = ...,
     coerce_float: bool = ...,
     params: list[str] | tuple[str, ...] | dict[str, str] | None = ...,

--- a/tests/test_io.py
+++ b/tests/test_io.py
@@ -42,6 +42,7 @@ from pandas import (
 from pandas._testing import ensure_clean
 import pytest
 import sqlalchemy
+import sqlalchemy.orm
 from typing_extensions import assert_type
 
 from tests import (
@@ -795,3 +796,16 @@ def test_csv_quoting():
             assert_type(DF.to_csv(path, quoting=csv.QUOTE_NONNUMERIC), None), type(None)
         )
         check(assert_type(DF.to_csv(path, quoting=csv.QUOTE_MINIMAL), None), type(None))
+
+
+def test_sqlalchemy_selectable() -> None:
+    with ensure_clean() as path:
+        db_uri = "sqlite:///" + path
+        engine = sqlalchemy.create_engine(db_uri)
+        with engine.connect():
+            check(assert_type(DF.to_sql("test", con=engine), Union[int, None]), int)
+
+        Session = sqlalchemy.orm.sessionmaker(engine)
+        with Session() as session:
+            if TYPE_CHECKING:
+                pd.read_sql(session.query("Something").statement, session.connection())

--- a/tests/test_io.py
+++ b/tests/test_io.py
@@ -42,7 +42,9 @@ from pandas import (
 from pandas._testing import ensure_clean
 import pytest
 import sqlalchemy
+import sqlalchemy.ext.declarative
 import sqlalchemy.orm
+import sqlalchemy.orm.decl_api
 from typing_extensions import assert_type
 
 from tests import (
@@ -802,10 +804,18 @@ def test_sqlalchemy_selectable() -> None:
     with ensure_clean() as path:
         db_uri = "sqlite:///" + path
         engine = sqlalchemy.create_engine(db_uri)
-        with engine.connect():
-            check(assert_type(DF.to_sql("test", con=engine), Union[int, None]), int)
 
-        Session = sqlalchemy.orm.sessionmaker(engine)
-        with Session() as session:
-            if TYPE_CHECKING:
-                pd.read_sql(session.query("Something").statement, session.connection())
+        if TYPE_CHECKING:
+            # Just type checking since underlying dB does not exist
+            class Base(metaclass=sqlalchemy.orm.decl_api.DeclarativeMeta):
+                __abstract__ = True
+
+            class Temp(Base):
+                __tablename__ = "part"
+                quantity = sqlalchemy.Column(sqlalchemy.Integer)
+
+            Session = sqlalchemy.orm.sessionmaker(engine)
+            with Session() as session:
+                pd.read_sql(
+                    session.query(Temp.quantity).statement, session.connection()
+                )

